### PR TITLE
Make default sidecar scope services merge consistent with sidecar

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Now we merge services that are from SE and in same namespace.
And k8s service takes priority, we donot merge k8s services' ports before and after. 

I am not sure of any regression by this, open for feedback first

### CODE REVIEW

File: pilot/pkg/xds/proxy_dependencies.go [hunk 1]
```diff
@@ -60,9 +60,9 @@ func proxyDependentOnConfig(proxy *model.Proxy, config model.ConfigKey, push *mo
 	// Detailed config dependencies check.
 	switch proxy.Type {
 	case model.SidecarProxy:
-		if proxy.SidecarScope.DependsOnConfig(config) {
+		if proxy.SidecarScope.DependsOnConfig(config, push.Mesh.RootNamespace) {
```
Comment (howardjohn) [inline pilot/pkg/xds/proxy_dependencies.go:63]: just curious, why did we split this out? Not a big deal, just slightly preferred the old approach

**Review by howardjohn — COMMENTED** [review]
Overall LGTM, very nice cleanup. One small comment + want to let @ramaraochavali review

File: pilot/pkg/xds/proxy_dependencies.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/proxy_dependencies.go:63]: yeah, a small change just to reduce SidecarScope unnecessary field

File: pilot/pkg/model/sidecar.go [hunk 2]
```diff
@@ -908,3 +842,66 @@ func needsPortMatch(l *networking.IstioEgressListener) bool {
...
+	if foundSvc, found := servicesAdded[s.Hostname]; !found {
+		sc.services = append(sc.services, s)
+		servicesAdded[s.Hostname] = sidecarServiceIndex{s, len(sc.services) - 1}
+		sc.servicesByHostname[s.Hostname] = s
+	} else {
+		existing := foundSvc.svc
+		// We donot merge k8s service with any other services from other registries
+		if existing.Attributes.ServiceRegistry == provider.Kubernetes {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:864]: Actually it is from before, k8s service has priority over SE and no merge with them

https://github.com/istio/istio/blob/94fc84de3f856913685292561569d5bfde079222/pilot/pkg/model/sidecar.go#L243-L253

**Comment by hzxuzhonghu** [comment]
@ramaraochavali Hope you have a chance to take a look

File: pilot/pkg/model/sidecar.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ type SidecarScope struct {
 	// This is the namespace where the sidecar takes effect,
 	// maybe different from the ns where sidecar resides if sidecar is in root ns.
 	Namespace string
-	// The crd itself. Can be nil if we are constructing the default
+	// The cr itself. Can be nil if we are constructing the default
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar.go:108]: nit: Is this intentional?

File: pilot/pkg/model/sidecar.go [hunk 4]
```diff
@@ -908,3 +842,66 @@ func needsPortMatch(l *networking.IstioEgressListener) bool {
...
+		if existing.Attributes.ServiceRegistry == provider.Kubernetes {
+			return
+		}
+		// In some scenarios, there may be multiple Services defined for the same hostname due to ServiceEntry allowing
+		// arbitrary hostnames. In these cases, we want to pick the first Service, which is the oldest. This ensures
+		// newly created Services cannot take ownership unexpectedly.
+		// However, the Service is from Kubernetes it should take precedence over ones not. This prevents someone from
+		// "domain squatting" on the hostname before a Kubernetes Service is created.
+		if s.Attributes.ServiceRegistry == provider.Kubernetes {
+			log.Debugf("Service %s/%s from registry %s ignored by %s/%s/%s", existing.Attributes.Namespace, existing.Hostname, existing.Attributes.ServiceRegistry,
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar.go:871]: This debug log line is incorrect now?

File: pilot/pkg/model/sidecar.go (hunk 3 above)
Comment (zirain) [reply pilot/pkg/model/sidecar.go:108]: cr short for custom resource, so I think it's intentional and more right.

File: pilot/pkg/model/sidecar.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/sidecar.go:871]: what do you mean

Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go:871]: nvm.. we come to this only if `existing` is service entry

**Review by ramaraochavali — APPROVED** [review]

**Pull request merged**